test/integration/integration.cc

```diff
@@ -414,10 +414,25 @@ void BaseIntegrationTest::setUpstreamAddress(
   socket_address->set_port_value(fake_upstreams_[upstream_index]->localAddress()->ip()->port());
 }
 
+namespace {
+bool read_flag(bool* flag) { return *flag; }
+} // namespace
+
 void BaseIntegrationTest::registerTestServerPorts(const std::vector<std::string>& port_names) {
-  auto port_it = port_names.cbegin();
-  auto listeners = test_server_->server().listenerManager().listeners();
+  bool listeners_ready = false;
+  absl::Mutex l;
+  std::vector<std::reference_wrapper<Network::ListenerConfig>> listeners;
+  test_server_->server().dispatcher().post([this, &listeners, &listeners_ready, &l]() {
+    listeners = test_server_->server().listenerManager().listeners();
+    l.Lock();
+    listeners_ready = true;
+    l.Unlock();
+  });
+  l.LockWhen(absl::Condition(read_flag, &listeners_ready));
+  l.Unlock();
+
   auto listener_it = listeners.cbegin();
+  auto port_it = port_names.cbegin();
```

Comment:
Wondering if this is safe given the comment about call stack validity in https://github.com/envoyproxy/envoy/blob/master/include/envoy/server/listener_manager.h#L164.

Reply:
I agree the production should not violate the rules. 
But for an integration test it's probably good enough. 
Well, if the PR is not strong enough, the next step is to provide a callback and copy the needed info in master thread.